packages/button/src/Button.tsx

```diff
@@ -1,483 +1,182 @@
-  [Size.XSmall]: css`
-    padding: 0 8px;
-  `,
-
-  [Size.Small]: css`
-    padding: 0 10px;
-  `,
-
-  [Size.Normal]: css`
-    padding: 0 12px;
-  `,
-
-  [Size.Large]: css`
-    padding: 0 20px;
-  `,
-} as const;
-
-const glyphMargins: Record<Size, string> = {
-  [Size.XSmall]: css`
-    margin-right: 2px;
-  `,
-
-  [Size.Small]: css`
-    margin-right: 4px;
-  `,
-
-  [Size.Normal]: css`
-    margin-right: 5px;
-  `,
-
-  [Size.Large]: css`
-    margin-right: 8px;
-  `,
-} as const;
-
-const focusStyle = css`
-  &:focus {
-    text-decoration: none;
-
-    &:not(:disabled):before {
-      opacity: 1;
-    }
-  }
-`;
-
-const baseStyle = css`
-  position: relative;
-  // Establishes the root element as a new stacking context
-  // so that the z-index of the span within the button doesn't
-  // appear above other elements on the page that it shouldn't.
-  z-index: 0;
-  border-radius: 3px;
-  box-sizing: border-box;
-  cursor: pointer;
-  display: inline-flex;
-  align-items: stretch;
-  font-family: Akzidenz, 'Helvetica Neue', Helvetica, Arial, sans-serif;
-  font-weight: normal;
-  text-decoration: none;
-  text-transform: none;
-  transition: all 120ms ease;
-  user-select: none;
-  padding: 0;
-
-  &:focus {
-    outline: none;
-  }
-
-  &:hover {
-    text-decoration: none;
-  }
-
-  // We're using CSS pseudo elements here in order to
-  // transition the gradients between button pseudo classes.
-  &:before,
-  &:after {
-    content: '';
-    transition: 0.15s opacity ease-in-out;
-    opacity: 0;
-    position: absolute;
-    top: 0;
-    bottom: 0;
-    left: 0;
-    right: 0;
-    border-radius: 2px;
-  }
-
-  &:not(:disabled) {
-    &:focus,
-    &:hover {
-      &:before {
-        opacity: 1;
-      }
-    }
+const iconSpacing = {
+  [Size.XSmall]: 6,
+  [Size.Small]: 6,
+  [Size.Default]: 6,
+  [Size.Large]: spacing[2],
+};
 
-    &:active:after {
-      opacity: 1;
-    }
+const Button: ExtendableBox<
+  ButtonProps & { ref?: React.Ref<any> },
+  'button'
+> = React.forwardRef(function Button(
+  {
+    variant = Variant.Default,
+    size = Size.Default,
+    darkMode = false,
+    baseFontSize = 14,
+    disabled = false,
+    leftGlyph,
+    rightGlyph,
+    children,
+    className,
+    ...rest
+  }: ButtonProps,
+  forwardRef,
+) {
+  const { usingKeyboard: showFocus } = useUsingKeyboardContext();
+  const rippleRef = useRef<HTMLDivElement | null>(null);
+
+  useEffect(() => {
+    if (rippleRef.current != null) {
+      registerRipple(rippleRef.current, { variant, darkMode });
+    }
+  }, [rippleRef, variant, darkMode]);
+
+  const isIconOnlyButton = ((leftGlyph || rightGlyph) && !children) ?? false;
+
+  const buttonClassName = getClassName({
+    variant,
+    size,
+    darkMode,
+    baseFontSize,
+    disabled,
+    showFocus,
+  });
+
+  const isAnchor = typeof rest.href === 'string';
+  let type: JSX.IntrinsicElements['button']['type'];
+
+  // @ts-expect-error rest.as may be defined
```

Comment:
I think the note here meant to say "undefined"

Also, I'm inclined to think we can address this TS error though code by checking whether rest.as is defined before the other checks below.

Reply:
Can we address this issue through JavaScript? I want to make sure we're taking these TS errors seriously so we're handling edge cases we may not have thought of.

Reply:
Not really.. this has more to do with our typing of box, but if we want to be more explicit with the component's internals we'll still need an expect-error directive when setting as on the Box component itself.

Reply:
Huh... Alright, I don't think we need to shave any yaks here, just usually these errors are resolvable by checking if `rest.as` exists before statements that would rely on that.